File: languages/lua/test_fatigue.lua
-- test_fatigue.lua
-- Busted test suite for Lua FatigueCalculator implementation
-- Run: busted test_fatigue.lua

local FC = require 'FatigueCalculator'

describe("FatigueCalculator", function()
    local calculator
    local default_profile = {
        sedentary_limit = 5.0,
        active_limit = 8.0,
        high_intensity_limit = 10.0
    }

    before_each(function()
        calculator = FC.new(default_profile)
    end)

    describe("Profile", function()
        it("should have correct default values", function()
            assert.are.equal(5.0, default_profile.sedentary_limit)
            assert.are.equal(8.0, default_profile.active_limit)
            assert.are.equal(10.0, default_profile.high_intensity_limit)
        end)

        it("get_limit should return correct limit for each activity level", function()
            assert.are.equal(5.0, calculator:get_limit(FC.ActivityLevel.SEDENTARY))
            assert.are.equal(8.0, calculator:get_limit(FC.ActivityLevel.ACTIVE))
            assert.are.equal(10.0, calculator:get_limit(FC.ActivityLevel.HIGH_INTENSITY))
        end)
    end)

    describe("calculate_fatigue", function()
        it("should return correct score for sedentary", function()
            local result = calculator:calculate_fatigue(3.0, FC.ActivityLevel.SEDENTARY)
            assert.is_true(result.success)
            assert.are.equal(3.0, result.value)
        end)

        it("should return correct score for active", function()
            local result = calculator:calculate_fatigue(6.0, FC.ActivityLevel.ACTIVE)
            assert.is_true(result.success)
            assert.are.equal(6.0, result.value)
        end)

        it("should return correct score for high_intensity", function()
            local result = calculator:calculate_fatigue(9.0, FC.ActivityLevel.HIGH_INTENSITY)
            assert.is_true(result.success)
            assert.are.equal(9.0, result.value)
        end)

        it("should clamp to limit", function()
            local result = calculator:calculate_fatigue(15.0, FC.ActivityLevel.SEDENTARY)
            assert.is_true(result.success)
            assert.are.equal(5.0, result.value) -- Clamped to sedentary limit
        end)

        it("should fail with negative score", function()
            local result = calculator:calculate_fatigue(-1.0, FC.ActivityLevel.SEDENTARY)
            assert.is_false(result.success)
            assert.are.equal(FC.Errors.NEGATIVE_SCORE, result.error)
        end)

        it("should handle zero score", function()
            local result = calculator:calculate_fatigue(0.0, FC.ActivityLevel.SEDENTARY)
            assert.is_true(result.success)
            assert.are.equal(0.0, result.value)
        end)
    end)

    describe("calculate_recovery", function()
        it("should reduce fatigue based on sleep", function()
            local result = calculator:calculate_recovery(10.0, 8.0)
            assert.is_true(result.success)
            assert.are.equal(9.2, result.value) -- 10 - (8 * 0.1)
        end)

        it("should clamp to zero", function()
            local result = calculator:calculate_recovery(0.5, 10.0)
            assert.is_true(result.success)
            assert.are.equal(0.0, result.value)
        end)

        it("should fail with negative fatigue", function()
            local result = calculator:calculate_recovery(-1.0, 8.0)
            assert.is_false(result.success)
        end)

        it("should fail with negative sleep", function()
            local result = calculator:calculate_recovery(10.0, -1.0)
            assert.is_false(result.success)
        end)

        it("should handle zero fatigue", function()
            local result = calculator:calculate_recovery(0.0, 8.0)
            assert.is_true(result.success)
            assert.are.equal(0.0, result.value)
        end)
    end)

    describe("summarize", function()
        it("should calculate correct statistics", function()
            local scores = {3.0, 5.0, 7.0, 9.0, 11.0}
            local result = calculator:summarize(scores)
            
            assert.is_true(result.success)
            assert.are.equal(35.0, result.value.total)
            assert.are.equal(5, result.value.count)
            assert.are.equal(7.0, result.value.average)
            assert.are.equal(11.0, result.value.maximum)
            assert.are.equal(3.0, result.value.minimum)
            assert.is_false(result.value.recovered) -- avg >= 5
        end)

        it("should mark recovered when average < 5", function()
            local scores = {1.0, 2.0, 3.0, 4.0}
            local result = calculator:summarize(scores)
            
            assert.is_true(result.success)
            assert.is_true(result.value.recovered)
        end)

        it("should fail with empty list", function()
            local result = calculator:summarize({})
            assert.is_false(result.success)
            assert.are.equal(FC.Errors.EMPTY_SLICE, result.error)
        end)

        it("should handle single value", function()
            local result = calculator:summarize({5.0})
            assert.is_true(result.success)
            assert.are.equal(5.0, result.value.total)
            assert.are.equal(1, result.value.count)
        end)
    end)

    describe("compute_intensity_metrics", function()
        it("should return correct value", function()
            local result = calculator:compute_intensity_metrics(10.0, 1.5)
            assert.is_true(result.success)
            assert.are.equal(15.0, result.value)
        end)

        it("should fail with negative multiplier", function()
            local result = calculator:compute_intensity_metrics(10.0, -1.0)
            assert.is_false(result.success)
        end)

        it("should handle zero multiplier", function()
            local result = calculator:compute_intensity_metrics(10.0, 0.0)
            assert.is_true(result.success)
            assert.are.equal(0.0, result.value)
        end)
    end)

    describe("adjust_for_activity_level", function()
        it("should adjust by sedentary factor", function()
            local result = calculator:adjust_for_activity_level(10.0, FC.ActivityLevel.SEDENTARY)
            -- 10 * 1.0 = 10, clamped to sedentary limit 5
            assert.is_true(result.success)
            assert.are.equal(5.0, result.value)
        end)

        it("should adjust by active factor", function()
            local result = calculator:adjust_for_activity_level(5.0, FC.ActivityLevel.ACTIVE)
            -- 5 * 1.2 = 6, within active limit 8
            assert.is_true(result.success)
            assert.are.equal(6.0, result.value)
        end)

        it("should adjust by high_intensity factor", function()
            local result = calculator:adjust_for_activity_level(5.0, FC.ActivityLevel.HIGH_INTENSITY)
            -- 5 * 1.5 = 7.5, within limit 10
            assert.is_true(result.success)
            assert.are.equal(7.5, result.value)
        end)
    end)

    describe("verify_threshold", function()
        it("should return true when computed exceeds threshold", function()
            local result = calculator:verify_threshold(10.0, 5.0)
            assert.is_true(result.success)
            assert.is_true(result.value)
        end)

        it("should return false when computed below threshold", function()
            local result = calculator:verify_threshold(4.0, 5.0)
            assert.is_true(result.success)
            assert.is_false(result.value)
        end)

        it("should fail with negative computed", function()
            local result = calculator:verify_threshold(-1.0, 5.0)
            assert.is_false(result.success)
        end)

        it("should fail with negative threshold", function()
            local result = calculator:verify_threshold(10.0, -1.0)
            assert.is_false(result.success)
        end)
    end)

    describe("categorize_fatigue_level", function()
        it("should return mild for score < 3", function()
            local result = calculator:categorize_fatigue_level(2.0)
            assert.is_true(result.success)
            assert.are.equal(FC.FatigueLevel.MILD, result.value)
        end)

        it("should return moderate for score 3-7", function()
            local result = calculator:categorize_fatigue_level(5.0)
            assert.is_true(result.success)
            assert.are.equal(FC.FatigueLevel.MODERATE, result.value)
        end)

        it("should return severe for score >= 7", function()
            local result = calculator:categorize_fatigue_level(8.0)
            assert.is_true(result.success)
            assert.are.equal(FC.FatigueLevel.SEVERE, result.value)
        end)

        it("should fail with negative score", function()
            local result = calculator:categorize_fatigue_level(-1.0)
            assert.is_false(result.success)
        end)
    end)

    describe("compute_z_score", function()
        it("should return correct z-score", function()
            local values = {1.0, 2.0, 3.0, 4.0, 5.0}
            local result = calculator:compute_z_score(values)
            
            assert.is_true(result.success)
            -- First value is below mean
            assert.is_true(result.value < 0)
        end)

        it("should fail with empty list", function()
            local result = calculator:compute_z_score({})
            assert.is_false(result.success)
            assert.are.equal(FC.Errors.EMPTY_SLICE, result.error)
        end)
    end)

    describe("calculate_rmse", function()
        it("should return correct rmse", function()
            local predicted = {1.0, 2.0, 3.0, 4.0, 5.0}
            local actual = {1.1, 1.9, 3.2, 3.8, 5.1}
            local result = calculator:calculate_rmse(predicted, actual)
            
            assert.is_true(result.success)
            assert.is_true(result.value > 0)
            assert.is_true(result.value < 0.5)
        end)

        it("should fail with mismatched arrays", function()
            local predicted = {1.0, 2.0, 3.0}
            local actual = {1.0, 2.0}
            local result = calculator:calculate_rmse(predicted, actual)
            assert.is_false(result.success)
        end)

        it("should fail with empty arrays", function()
            local result = calculator:calculate_rmse({}, {})
            assert.is_false(result.success)
        end)
    end)

    describe("generate_forecast_trend", function()
        it("should return correct forecasts", function()
            local historical = {10.0, 11.0, 12.0, 13.0}
            local result = calculator:generate_forecast_trend(historical, 3)
            
            assert.is_true(result.success)
            assert.are.equal(3, #result.value)
            assert.are.equal(14.0, result.value[1])
            assert.are.equal(15.0, result.value[2])
            assert.are.equal(16.0, result.value[3])
        end)

        it("should fail with empty historical", function()
            local result = calculator:generate_forecast_trend({}, 3)
            assert.is_false(result.success)
        end)

        it("should fail with zero periods", function()
            local result = calculator:generate_forecast_trend({10.0}, 0)
            assert.is_false(result.success)
        end)
    end)

    describe("empty_summary", function()
        it("should return empty summary correctly", function()
            local summary = FC.empty_summary()
            assert.are.equal(0, summary.total)
            assert.are.equal(0, summary.count)
            assert.is_false(summary.recovered)
        end)
    end)
end)

File: languages/zig/test_fatigue.zig
// test_fatigue.zig
// Zig test suite for fatigue calculation algorithms

const std = @import("std");
const testing = std.testing;
const fatigue = @import("fatigue_calculator.zig");

const FatigueCalculator = fatigue.FatigueCalculator;
const FatigueProfile = fatigue.FatigueProfile;
const ActivityLevel = fatigue.ActivityLevel;
const FatigueError = fatigue.FatigueError;
const FatigueSummary = fatigue.FatigueSummary;

// Test allocator for memory management
var gpa = std.heap.GeneralPurposeAllocator(.{}){};
const allocator = gpa.allocator();

// Default profile for tests
const default_profile = FatigueProfile{
    .sedentary_limit = 5.0,
    .active_limit = 8.0,
    .high_intensity_limit = 10.0,
};

// Profile Tests
test "profile should have correct default values" {
    try testing.expect(default_profile.sedentary_limit == 5.0);
    try testing.expect(default_profile.active_limit == 8.0);
    try testing.expect(default_profile.high_intensity_limit == 10.0);
}

test "getLimit should return correct limit for each activity level" {
    // These use internal getLimit which is tested via calculateFatigue
    const sed_limit = default_profile.getLimit(ActivityLevel.sedentary);
    const act_limit = default_profile.getLimit(ActivityLevel.active);
    const high_limit = default_profile.getLimit(ActivityLevel.high_intensity);
    
    try testing.expect(sed_limit == 5.0);
    try testing.expect(act_limit == 8.0);
    try testing.expect(high_limit == 10.0);
}

// Initialization Tests
test "calculator should initialize with default profile" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    // Just verify it initializes without error
    _ = calc.profile;
}

// calculateFatigue Tests
test "calculateFatigue should return correct score for sedentary" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateFatigue(3.0, ActivityLevel.sedentary);
    try testing.expect(result == 3.0);
}

test "calculateFatigue should return correct score for active" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateFatigue(6.0, ActivityLevel.active);
    try testing.expect(result == 6.0);
}

test "calculateFatigue should return correct score for high intensity" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateFatigue(9.0, ActivityLevel.high_intensity);
    try testing.expect(result == 9.0);
}

test "calculateFatigue should clamp to limit" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateFatigue(15.0, ActivityLevel.sedentary);
    try testing.expect(result == 5.0); // Should clamp to sedentary limit
}

test "calculateFatigue should fail with negative score" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.calculateFatigue(-1.0, ActivityLevel.sedentary);
    try testing.expectError(FatigueError.NegativeScore, result);
}

test "calculateFatigue should handle zero score" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateFatigue(0.0, ActivityLevel.sedentary);
    try testing.expect(result == 0.0);
}

// calculateRecovery Tests
test "calculateRecovery should reduce fatigue based on sleep" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateRecovery(10.0, 8.0);
    try testing.expect(result == 9.2); // 10 - (8 * 0.1)
}

test "calculateRecovery should clamp to zero" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateRecovery(0.5, 10.0);
    try testing.expect(result == 0.0);
}

test "calculateRecovery should fail with negative fatigue" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.calculateRecovery(-1.0, 8.0);
    try testing.expectError(FatigueError.NegativeScore, result);
}

test "calculateRecovery should fail with negative sleep" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.calculateRecovery(10.0, -1.0);
    try testing.expectError(FatigueError.NegativeScore, result);
}

test "calculateRecovery should handle zero fatigue" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.calculateRecovery(0.0, 8.0);
    try testing.expect(result == 0.0);
}

// summarize Tests
test "summarize should calculate correct statistics" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const scores = [_]f64{ 3.0, 5.0, 7.0, 9.0, 11.0 };
    const summary = try calc.summarize(&scores);
    
    try testing.expect(summary.total == 35.0);
    try testing.expect(summary.count == 5);
    try testing.expect(summary.average == 7.0);
    try testing.expect(summary.maximum == 11.0);
    try testing.expect(summary.minimum == 3.0);
    try testing.expect(!summary.recovered); // avg >= 5
}

test "summarize should mark recovered when average below threshold" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const scores = [_]f64{ 1.0, 2.0, 3.0, 4.0 };
    const summary = try calc.summarize(&scores);
    
    try testing.expect(summary.recovered); // avg < 5
}

test "summarize should fail with empty slice" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const scores = [_]f64{};
    const result = calc.summarize(&scores);
    try testing.expectError(FatigueError.EmptySlice, result);
}

test "summarize should handle single value" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const scores = [_]f64{5.0};
    const summary = try calc.summarize(&scores);
    
    try testing.expect(summary.total == 5.0);
    try testing.expect(summary.count == 1);
    try testing.expect(summary.maximum == 5.0);
    try testing.expect(summary.minimum == 5.0);
}

// computeIntensityMetrics Tests
test "computeIntensityMetrics should return correct value" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.computeIntensityMetrics(10.0, 1.5);
    try testing.expect(result == 15.0);
}

test "computeIntensityMetrics should fail with negative multiplier" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.computeIntensityMetrics(10.0, -1.0);
    try testing.expectError(FatigueError.InvalidLevel, result);
}

test "computeIntensityMetrics should handle zero multiplier" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.computeIntensityMetrics(10.0, 0.0);
    try testing.expect(result == 0.0);
}

// adjustForActivityLevel Tests
test "adjustForActivityLevel should adjust by sedentary factor" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.adjustForActivityLevel(10.0, ActivityLevel.sedentary);
    try testing.expect(result == 5.0); // Clamped to limit
}

test "adjustForActivityLevel should adjust by active factor" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.adjustForActivityLevel(5.0, ActivityLevel.active);
    try testing.expect(result == 6.0); // 5 * 1.2 = 6
}

test "adjustForActivityLevel should adjust by high intensity factor" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.adjustForActivityLevel(5.0, ActivityLevel.high_intensity);
    try testing.expect(result == 7.5); // 5 * 1.5 = 7.5
}

// verifyThreshold Tests
test "verifyThreshold should return true when computed exceeds threshold" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.verifyThreshold(10.0, 5.0);
    try testing.expect(result == true);
}

test "verifyThreshold should return false when computed below threshold" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = try calc.verifyThreshold(4.0, 5.0);
    try testing.expect(result == false);
}

test "verifyThreshold should fail with negative computed" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.verifyThreshold(-1.0, 5.0);
    try testing.expectError(FatigueError.NegativeScore, result);
}

test "verifyThreshold should fail with negative threshold" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.verifyThreshold(10.0, -1.0);
    try testing.expectError(FatigueError.NegativeScore, result);
}

// Other Algorithm Tests
test "categorizeFatigueLevel should return mild" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const level = try calc.categorizeFatigueLevel(2.0);
    try testing.expect(level == fatigue.FatigueLevel.mild);
}

test "categorizeFatigueLevel should return moderate" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const level = try calc.categorizeFatigueLevel(5.0);
    try testing.expect(level == fatigue.FatigueLevel.moderate);
}

test "categorizeFatigueLevel should return severe" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const level = try calc.categorizeFatigueLevel(8.0);
    try testing.expect(level == fatigue.FatigueLevel.severe);
}

test "categorizeFatigueLevel should fail with negative score" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const result = calc.categorizeFatigueLevel(-1.0);
    try testing.expectError(FatigueError.NegativeScore, result);
}

test "computeZScore should return correct value" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const values = [_]f64{ 1.0, 2.0, 3.0, 4.0, 5.0 };
    const zScore = try calc.computeZScore(&values);
    
    // First value is below mean, so z-score should be < 0
    try testing.expect(zScore < 0);
}

test "computeZScore should fail with empty slice" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const values = [_]f64{};
    const result = calc.computeZScore(&values);
    try testing.expectError(FatigueError.EmptySlice, result);
}

// calculateRMSE Tests
test "calculateRMSE should return correct value" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const predicted = [_]f64{ 1.0, 2.0, 3.0, 4.0, 5.0 };
    const actual = [_]f64{ 1.1, 1.9, 3.2, 3.8, 5.1 };
    const rmse = try calc.calculateRMSE(&predicted, &actual);
    
    try testing.expect(rmse > 0);
    try testing.expect(rmse < 0.5); // Should be small
}

test "calculateRMSE should fail with mismatched arrays" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const predicted = [_]f64{ 1.0, 2.0, 3.0 };
    const actual = [_]f64{ 1.0, 2.0 };
    const result = calc.calculateRMSE(&predicted, &actual);
    try testing.expectError(FatigueError.EmptySlice, result);
}

test "calculateRMSE should fail with empty arrays" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const predicted = [_]f64{};
    const actual = [_]f64{};
    const result = calc.calculateRMSE(&predicted, &actual);
    try testing.expectError(FatigueError.EmptySlice, result);
}

// generateForecastTrend Tests
test "generateForecastTrend should return correct forecasts" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const historical = [_]f64{ 10.0, 11.0, 12.0, 13.0 };
    const forecasts = try calc.generateForecastTrend(&historical, 3);
    defer allocator.free(forecasts);
    
    try testing.expect(forecasts.len == 3);
    try testing.expect(forecasts[0] == 14.0);
    try testing.expect(forecasts[1] == 15.0);
    try testing.expect(forecasts[2] == 16.0);
}

test "generateForecastTrend should fail with empty historical" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const historical = [_]f64{};
    const result = calc.generateForecastTrend(&historical, 3);
    try testing.expectError(FatigueError.EmptySlice, result);
}

test "generateForecastTrend should fail with zero periods" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    const historical = [_]f64{ 10.0 };
    const result = calc.generateForecastTrend(&historical, 0);
    try testing.expectError(FatigueError.EmptySlice, result);
}

// Memory safety test
test "calculator should handle memory allocation" {
    var calc = FatigueCalculator.init(allocator, default_profile);
    defer calc.deinit();
    
    // This test ensures proper memory cleanup
    const historical = [_]f64{ 10.0, 11.0, 12.0 };
    const forecasts = try calc.generateForecastTrend(&historical, 5);
    defer allocator.free(forecasts);
    
    try testing.expect(forecasts.len == 5);
}
